db/migrate/20170821152307_add_source_type_to_permission_templates.rb

```diff
@@ -1,7 +1,16 @@
 class AddSourceTypeToPermissionTemplates < ActiveRecord::Migration[4.2]
   def change
     add_column :permission_templates, :source_type, :string
-    rename_column :permission_templates, :admin_set_id, :source_id
+    add_column :permission_templates, :source_id, :string
+
+    Hyrax::PermissionTemplate.find_each do |permission_template|
+      permission_template.source_type = 'admin_set'
+      permission_template.source_id = permission_template.admin_set_id
+      permission_template.save!
+    end
+
+    # If the db is sqlite, don't drop the column or you will get a FK error
+    # because of limitation in sqlite
+    remove_column :permission_templates, :admin_set_id, :string unless connection.adapter_name.downcase.starts_with?('sqlite')
```

Comment:
Not keen on conditional migrations, because the SQLite database will keep the admin_set_id (and the foreign key constraint).

Take a look at http://api.rubyonrails.org/v4.2/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-remove_foreign_key for removing foreign key constraints. I believe you'll need to first remove the constraint, then remove the column.

Reply:
In Sqlite, the FK constraint does not get removed.  In other words, `remove_foreign_key` has no affect in Sqlite.

Reply:
It seems like this will be (or should be?) such a rare case that it's reasonable to expect manual intervention to migrate the table themselves. If so, I would suggest just removing the conditional. Let it fail when this happens. Perhaps output some text about this being a known issue and what to do when it happens. 

To migrate it themselves, they would have to:
- Recreate permission_template_accesses without the 't.references :permission_template, foreign_key: true'
- Run the migration
- Recreate permission_template_accesses with the 't.references :permission_template, foreign_key: true'

All because sqlite doesn't support dropping FKC in an ALTER TABLE statement :(

That, or do a conditional migration that does all of this...

Reply:
If you need to see an example of the migration I'm describing above, I have a working one here https://github.com/samvera/hyrax/compare/collections-sprint...coll-1580-sqlite-fix. My suggestion is still to not do this, and just fix the data problem. I've created a PR that focuses on only the data https://github.com/samvera/hyrax/pull/2310. If that one is accepted, perhaps we can just close this one.

Reply:
@mjgiarlo @vantuyls I agree with @jgondron's accessment of the migration script.  There was one missing piece from the original migration script that does need to be addressed which left the source_type blank instead of setting it to 'admin_set'.  That is remedied by PR #2310.

So the proposal is to close this PR without merging, merge PR #2310, and provide migration instructions that explain the problem with sqlite.  

The instructions can include an explanation that sqlite is not recommended for production systems.  If they are using sqlite for development and need to migrate, then follow the instructions as outlined by @jgondron above.